File: demotivators/demotivator.js
const { createWriteStream } = require('fs');
const { promisify } = require('util');
const { join } = require('path');
const { default: axios } = require('axios');
const { createCanvas, loadImage, Image, registerFont } = require('canvas');
const { expand } = require('canvas-constructor/lib/Util');
const { text, drawImage } = require('canvas-constructor');

class Demotivator {
  constructor(topText = '', bottomText = '') {
    this._topText = topText;
    this._bottomText = bottomText;
  }

  async create(
    file,
    watermark = null,
    resultFilename = 'demresult.jpg',
    fontColor = 'white',
    fillColor = 'black',
    fontName = 'times.ttf',
    topSize = 80,
    bottomSize = 60,
    arrange = false,
    useUrl = false,
    deleteFile = false,
  ) {
    if (useUrl) {
      const { data } = await axios.get(file, { responseType: 'arraybuffer' });
      file = join(__dirname, 'demotivator_picture.jpg');
      await promisify(createWriteStream(file))(Buffer.from(data, 'binary'));
    }

    /*
      Создаем шаблон для демотиватора
      Вставляем фотографию в рамку
    */

    let userImg = await loadImage(file);
    let width, height;
    if (arrange) {
      ({ width, height } = userImg);
      width += 250;
      height += 260;
    } else {
      width = 1280;
      height = 1024;
    }
    const img = createCanvas(width, height);
    const imgBorder = createCanvas(width - 10, height - 10);
    const border = expand(imgBorder, 2, '#ffffff');
    const drawer = img.getContext('2d');

    if (arrange) {
      drawer.fillStyle = fillColor;
      drawer.fillRect(0, 0, width, height);
      drawer.drawImage(border, 111, 96);
      drawer.drawImage(userImg, 118, 103);
    } else {
      drawer.fillStyle = fillColor;
      drawer.fillRect(0, 0, width, height);
      drawer.drawImage(border, 111, 96);
      drawer.drawImage(userImg, 118, 103, 1050, 710);
    }

    /*
      Подбираем оптимальный размер шрифта
      Добавляем текст в шаблон для демотиватора
    */

    registerFont(join(__dirname, fontName), { family: 'Custom' });
    let font1 = `${topSize}px Custom`;
    let textWidth = drawer.measureText(this._topText).width;
    while (textWidth >= width - 20) {
      font1 = `${--topSize}px Custom`;
      drawer.font = font1;
      textWidth = drawer.measureText(this._topText).width;
    }
    let font2 = `${bottomSize}px Custom`;
    textWidth = drawer.measureText(this._bottomText).width;
    while (textWidth >= width - 20) {
      font2 = `${--bottomSize}px Custom`;
      drawer.font = font2;
      textWidth = drawer.measureText(this._bottomText).width;
    }
    drawer.fillStyle = fontColor;
    drawer.font = font1;
    const size1 = [drawer.measureText(this._topText).width, topSize];
    drawer.fillText(this._topText, (width - size1) / 2, 80);
    drawer.font = font2;
    const size2 = [drawer.measureText(this._bottomText).width, bottomSize];
    drawer.fillText(this._bottomText, (width - size2[0]) / 2, height - 50);
    /*
    Добавляем водяной знак, если он указан
    Сохраняем результат в файл
    */

    if (watermark) {
        const watermarkImg = await loadImage(watermark);
        drawer.drawImage(watermarkImg, 20, 20, 80, 80);
    }
    
    const stream = createWriteStream(resultFilename);
    const out = img.createJPEGStream({
        quality: 0.9,
        chromaSubsampling: true,
    });
    
    out.pipe(stream);
    await new Promise((resolve) => {
        out.on('end', resolve);
    });
    
    if (deleteFile && useUrl) {
        await promisify(unlink)(file);
    }
    
    return resultFilename;}
}

module.exports = Demotivator;
    

File: app/simpledemotivators_2_0/demotivator.py
import os
from PIL import Image, ImageDraw, ImageFont, ImageOps
import aiohttp
import aiofiles
import asyncio

class Demotivator:
    def __init__(self, top_text='', bottom_text=''):
        self._top_text = top_text
        self._bottom_text = bottom_text

    async def _download_image(self, url, local_path):
        """
        Загружает изображение асинхронно.
        """
        async with aiohttp.ClientSession() as session:
            async with session.get(url) as response:
                if response.status == 200:
                    async with aiofiles.open(local_path, 'wb') as out_file:
                        await out_file.write(await response.read())

    async def create(self, folder_name: str, avatar_name: str, result_filename: str, watermark=None, font_color='white', 
                     fill_color='black', font_name='Impact.ttf', top_size=80, bottom_size=60,
                     arrange=False, use_url=False) -> bool:
        """
        Создает демотиватор с именем файла, основанным на переданном параметре `result_filename`.
        
        :param folder_name: Папка для сохранения файла и загрузки изображения.
        :param avatar_name: Имя изображения, которое будет использовано.
        :param result_filename: Имя итогового файла.
        :param watermark: Текст водяного знака, если None, то не будет добавлен.
        :param font_color: Цвет шрифта.
        :param fill_color: Цвет фона.
        :param font_name: Путь к файлу шрифта.
        :param top_size: Размер шрифта для верхнего текста.
        :param bottom_size: Размер шрифта для нижнего текста.
        :param arrange: True, если фотография должна быть вставлена в рамку.
        :param use_url: True, если `avatar_name` - это URL.
        :return: True, если метод выполнен успешно.
        """
        # Путь к изображению (если это локальный файл или URL)
        avatar_path = os.path.join(folder_name, avatar_name)

        # Если используется URL, скачиваем изображение
        if use_url:
            local_file = os.path.join(folder_name, 'downloaded_image.jpg')
            await self._download_image(avatar_name, local_file)
            avatar_path = local_file  # Теперь используем локальный путь

        # Загружаем изображение
        user_img = Image.open(avatar_path).convert("RGBA")
        (width, height) = user_img.size
        
        # Создаем изображение для демотиватора с рамкой
        if arrange:
            img = Image.new('RGB', (width + 250, height + 260), color=fill_color)
            img_border = Image.new('RGB', (width + 10, height + 10), color='#000000')
            border = ImageOps.expand(img_border, border=2, fill='#ffffff')
            img.paste(border, (111, 96))
            img.paste(user_img, (118, 103))
            drawer = ImageDraw.Draw(img)
        else:
            img = Image.new('RGB', (1280, 1024), color=fill_color)
            img_border = Image.new('RGB', (1060, 720), color='#000000')
            border = ImageOps.expand(img_border, border=2, fill='#ffffff')
            user_img = user_img.resize((1050, 710))
            (width, height) = user_img.size
            img.paste(border, (111, 96))
            img.paste(user_img, (118, 103))
            drawer = ImageDraw.Draw(img)

        # Работа с шрифтами и текстом
        font_1 = ImageFont.truetype(font=os.path.join(os.path.dirname(__file__), font_name), size=top_size, encoding='UTF-8')
        text_bbox = font_1.getbbox(self._top_text)
        text_width = text_bbox[2] - text_bbox[0]

        while text_width >= (width + 250) - 20:
            font_1 = ImageFont.truetype(font=os.path.join(os.path.dirname(__file__), font_name), size=top_size, encoding='UTF-8')
            text_bbox = font_1.getbbox(self._top_text)
            text_width = text_bbox[2] - text_bbox[0]
            top_size -= 1

        font_2 = ImageFont.truetype(font=os.path.join(os.path.dirname(__file__), font_name), size=bottom_size, encoding='UTF-8')
        text_bbox = font_2.getbbox(self._bottom_text)
        text_width = text_bbox[2] - text_bbox[0]

        while text_width >= (width + 250) - 20:
            font_2 = ImageFont.truetype(font=os.path.join(os.path.dirname(__file__), font_name), size=bottom_size, encoding='UTF-8')
            text_bbox = font_2.getbbox(self._bottom_text)
            text_width = text_bbox[2] - text_bbox[0]
            bottom_size -= 1

        size_1_bbox = drawer.textbbox((0, 0), self._top_text, font=font_1)
        size_1 = (size_1_bbox[2] - size_1_bbox[0], size_1_bbox[3] - size_1_bbox[1])
        
        size_2_bbox = drawer.textbbox((0, 0), self._bottom_text, font=font_2)
        size_2 = (size_2_bbox[2] - size_2_bbox[0], size_2_bbox[3] - size_2_bbox[1])

        if arrange:
            drawer.text((((width + 250) - size_1[0]) / 2, ((height + 190) - size_1[1])),
                        self._top_text, fill=font_color,
                        font=font_1)
            drawer.text((((width + 250) - size_2[0]) / 2, ((height + 235) - size_2[1])),
                        self._bottom_text, fill=font_color,
                        font=font_2)
        else:
            drawer.text(((1280 - size_1[0]) / 2, 840), self._top_text, fill=font_color, font=font_1)
            drawer.text(((1280 - size_2[0]) / 2, 930), self._bottom_text, fill=font_color, font=font_2)

        # Добавление водяного знака, если нужно
        if watermark is not None:
            (width, height) = img.size
            idraw = ImageDraw.Draw(img)
            idraw.line((1000 - len(watermark) * 5, 817, 1008 + len(watermark) * 5, 817), fill=0, width=4)

            font_2 = ImageFont.truetype(font=os.path.join(os.path.dirname(__file__), font_name), size=20, encoding='UTF-8')
            size_2_bbox = idraw.textbbox((0, 0), watermark.lower(), font=font_2)
            size_2 = (size_2_bbox[2] - size_2_bbox[0], size_2_bbox[3] - size_2_bbox[1])
            idraw.text((((width + 729) - size_2[0]) / 2, ((height - 192) - size_2[1])),
                       watermark.lower(), font=font_2)

        # Сохраняем результат
        result_filename = f"{result_filename}_dem.jpg"
        result_filepath = os.path.join(folder_name, result_filename)
        img.save(result_filepath)

        # Удаляем файл, если он был загружен по URL
        if use_url:
            os.remove(avatar_path)

        return True
